GH-43956: [C++][Format] Add initial Decimal32/Decimal64 implementations

zeroshade opened the pull request:
<!--
Thanks for opening a pull request!
If this is your first pull request you can find detailed information on how 
to contribute here:
  * [New Contributor's Guide](https://arrow.apache.org/docs/dev/developers/guide/step_by_step/pr_lifecycle.html#reviews-and-merge-of-the-pull-request)
  * [Contributing Overview](https://arrow.apache.org/docs/dev/developers/overview.html)


If this is not a [minor PR](https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#Minor-Fixes). Could you open an issue for this pull request on GitHub? https://github.com/apache/arrow/issues/new/choose

Opening GitHub issues ahead of time contributes to the [Openness](http://theapacheway.com/open/#:~:text=Openness%20allows%20new%20users%20the,must%20happen%20in%20the%20open.) of the Apache Arrow project.

Then could you also rename the pull request title in the following format?

    GH-${GITHUB_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

or

    MINOR: [${COMPONENT}] ${SUMMARY}

In the case of PARQUET issues on JIRA the title also supports:

    PARQUET-${JIRA_ISSUE_ID}: [${COMPONENT}] ${SUMMARY}

-->

### Rationale for this change
Widening the Decimal128/256 type to allow for bitwidths of 32 and 64 allows for more interoperability with other libraries and utilities which already support these types. This provides even more opportunities for zero-copy interactions between things such as libcudf and various databases.

<!--
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  
-->

### What changes are included in this PR?
This PR contains the basic C++ implementations for Decimal32/Decimal64 types, arrays, builders and scalars. It also includes the minimum necessary to get everything compiling and tests passing without also extending the acero kernels and parquet handling (both of which will be handled in follow-up PRs).

<!--
There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.
-->

### Are these changes tested?
Yes, tests were extended where applicable to add decimal32/decimal64 cases.

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

### Are there any user-facing changes?
Currently if a user is using `decimal(precision, scale)` rather than `decimal128(precision, scale)` they will get a `Decimal128Type` if the precision is <= 38 (max precision for Decimal128) and `Decimal256Type` if the precision is higher. Following the same pattern, this change means that using `decimal(precision, scale)` instead of the specific `decimal32`/`decimal64`/`decimal128`/`decimal256` functions results in the following functionality:

- for precisions [1 : 9] => `Decimal32Type`
- for precisions [10 : 18] => `Decimal64Type`
- for precisions [19 : 38] => `Decimal128Type`
- for precisions [39 : 76] => `Decimal256Type`

While many of our tests currently make the assumption that `decimal` with a low precision would be `Decimal128` and had to be updated, this may cause an initial surprise if users are making the same assumptions.

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please uncomment the line below and explain which changes are breaking.
-->
<!-- **This PR includes breaking changes to public APIs.** -->

<!--
Please uncomment the line below (and provide explanation) if the changes fix either (a) a security vulnerability, (b) a bug that caused incorrect or invalid data to be produced, or (c) a bug that causes a crash (even when the API contract is upheld). We use this to highlight fixes to issues that may affect users without their knowledge. For this reason, fixing bugs that cause errors don't count, since those are usually obvious.
-->
<!-- **This PR contains a "Critical Fix".** -->
* GitHub Issue: #43956

lidavidm commented on cpp/src/arrow/array/builder_dict.h:
```diff
@@ -298,6 +298,22 @@ class DictionaryBuilderBase : public ArrayBuilder {
     return Append(std::string_view(value, length));
   }
 
+  /// \brief Append a decimal (only for Decimal32Type)
```
We can do all this with `if constexpr` now right? (Though this works fine too)

lidavidm commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -1116,6 +1168,10 @@ ArrayKernelExec GeneratePhysicalNumeric(detail::GetTypeId get_id) {
 template <template <typename... Args> class Generator, typename... Args>
 ArrayKernelExec GenerateDecimalToDecimal(detail::GetTypeId get_id) {
   switch (get_id.id) {
+    // case Type::DECIMAL32:
+    //   return Generator<Decimal32Type, Args...>::Exec;
+    // case Type::DECIMAL64:
+    //   return Generator<Decimal64Type, Args...>::Exec;
```
(why are these commented out?)

lidavidm commented on cpp/src/arrow/testing/gtest_util.h:
```diff
@@ -171,7 +171,8 @@ using PrimitiveArrowTypes =
 
-using DecimalArrowTypes = ::testing::Types<Decimal128Type, Decimal256Type>;
+using DecimalArrowTypes =
+    ::testing::Types</*Decimal32Type, Decimal64Type,*/ Decimal128Type, Decimal256Type>;
```
Ditto here. (Should we file issues to come back to these?)

zeroshade replied:
These are commented out because we didn't implement casting for the new decimal types. This is mentioned in the issue as check boxes to do rather than as an entirely separate issue currently.

pitrou replied:
But it's going to be a separate PR, right?

pitrou commented:
> Following the same pattern, this change means that using `decimal(precision, scale)` instead of the specific `decimal32`/`decimal64`/`decimal128`/`decimal256` functions results in the following functionality

I'm afraid this may massively break user code. I would suggest another approach:
* deprecate the `decimal()` factory while keeping its current behavior of always returning at least `decimal128`
* introduce a new `smallest_decimal()` factory that is documented to return the smallest possible type, and explicitly makes no guarantees about the stability of the return type

wgtmac commented on cpp/src/arrow/type.cc:
```diff
@@ -3301,8 +3356,18 @@ std::shared_ptr<Field> field(std::string name, std::shared_ptr<DataType> type,
 std::shared_ptr<DataType> decimal(int32_t precision, int32_t scale) {
-  return precision <= Decimal128Type::kMaxPrecision ? decimal128(precision, scale)
-                                                    : decimal256(precision, scale);
+  return precision <= Decimal32Type::kMaxPrecision    ? decimal32(precision, scale)
```
Is it a breaking change? Users may get different data type and array type when precision is smaller than 18.

bkietz replied:
This is a breaking change, and personally I'd prefer we deprecate this function.

zeroshade replied:
the updated version does exactly that, added a deprecation notice and created a new function

zeroshade commented:
@pitrou @bkietz @wgtmac I've updated this based on the suggestion, created a `smallest_decimal` function and added a deprecated message to the docstring for `decimal`.

kou commented on cpp/src/arrow/c/bridge.cc:
```diff
@@ -1253,6 +1253,10 @@ struct SchemaImporter {
       type_ = decimal128(prec_scale[0], prec_scale[1]);
     } else if (prec_scale[2] == 256) {
       type_ = decimal256(prec_scale[0], prec_scale[1]);
+    } else if (prec_scale[2] == 32) {
+      type_ = decimal32(prec_scale[0], prec_scale[1]);
+    } else if (prec_scale[2] == 64) {
+      type_ = decimal64(prec_scale[0], prec_scale[1]);
```
Can we use 32 -> 64 -> 128 -> 256 order here?

kou commented on cpp/src/arrow/integration/json_internal.cc:
```diff
@@ -969,11 +1009,17 @@ Result<std::shared_ptr<DataType>> GetDecimal(const RjObject& json_type) {
+      return decimal256(precision, scale);
   }
+
   return Status::Invalid("Only 128 bit and 256 Decimals are supported. Received",
                          bit_width);
```
Could you update this message too?

felipecrv commented on cpp/src/arrow/ipc/metadata_internal.cc:
```diff
@@ -278,13 +278,20 @@ Status ConcreteTypeFromFlatbuffer(flatbuf::Type type, const void* type_data,
+              .Value(out);
       }
+      return Status::Invalid(
+          "Library supports 32-bit, 64-bit, 128-bit or 256-bit decimal values");
```
```suggestion
          "Library only supports 32/64/128/256-bit decimal values");
```

felipecrv reviewed (commented):
Initial comments [1]. It's a long PR.

[1] I reviewed until `ipc/`.

zeroshade commented on cpp/src/arrow/array/array_test.cc:
```diff
@@ -667,8 +667,10 @@ static ScalarVector GetScalars() {
+      std::make_shared<Decimal32Scalar>(Decimal32(10), smallest_decimal(7, 4)),
+      std::make_shared<Decimal64Scalar>(Decimal64(10), smallest_decimal(12, 4)),
+      std::make_shared<Decimal128Scalar>(Decimal128(10), smallest_decimal(20, 4)),
+      std::make_shared<Decimal256Scalar>(Decimal256(10), smallest_decimal(76, 38)),
```
`smallest_decimal` was the suggestion by @pitrou, no one made any objections and I doubt he's particularly tied to the name if you have a suggestion for a better one. What are your thoughts?

bkietz replied:
"narrowest_decimal"?

felipecrv replied:
I added suggestions here https://github.com/apache/arrow/pull/43957#discussion_r1752947006

felipecrv replied:
Alright. Can we name it something shorter? I have at least 3 suggestions:
- `best_decimal`
- `fast_decimal`
- `least_decimal`

Some of these naming conventions exist in the C++ standard library https://en.cppreference.com/w/cpp/types/integer

pitrou replied:
"least_decimal" is the best of these 3 suggestions ("best" and "fast" do not convey the right meaning), but "smallest" is easier to remember than "least" IMHO.

zeroshade replied:
I agree with @pitrou that "smallest" is easier to remember than "least" but I'm fine with renaming it to "least" if that is preferred.

@felipecrv, @bkietz, @lidavidm thoughts?

zeroshade commented on cpp/src/arrow/compute/kernels/aggregate_internal.h:
```diff
@@ -52,6 +52,16 @@ struct FindAccumulatorType<I, enable_if_floating_point<I>> {
 
+template <typename I>
+struct FindAccumulatorType<I, enable_if_decimal32<I>> {
+  using Type = Decimal32Type;
```
Makes sense, i'll fix that

zeroshade commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -140,6 +140,30 @@ struct GetViewType<Type, enable_if_t<is_base_binary_type<Type>::value ||
+template <>
+struct GetViewType<Decimal32Type> {
+  using T = Decimal32;
+  using PhysicalType = std::string_view;
```
what is the `PhysicalType` here used for compared to the `T` itself? (I was just following the existing pattern for `Decimal128/Decimal256` when I did this)

zeroshade replied:
Doing a little digging, the reason why we can't change it to `int32_t` here is because all of the decimal types are treated as `fixed_size_binary` rather than `numeric` in the rest of the codebase

zeroshade replied:
The issue comes down to `visit_data_inline.h` which uses `std::string_view` to call these `GetViewType` functions, so I don't think we could replace these `std::string_view` usages without significant refactors elsewhere.

felipecrv replied:
Yep. Some templates made the bad decision of overloading on the PhysicalType so integers can't be used as PhysicalType for other types that are not the numeric ones. This is specially problematic when trying to improve the support for half-float.

mapleFU replied:
I also think maybe int32 or `span<char, 4>` better here

zeroshade commented on cpp/src/arrow/compute/kernels/aggregate_var_std.cc:
```diff
@@ -46,15 +46,18 @@ struct VarStdState {
-  enable_if_t<is_floating_type<T>::value || (sizeof(CType) > 4)> Consume(
-      const ArraySpan& array) {
+  enable_if_t<is_floating_type<T>::value || (sizeof(CType) > 4) ||
+              (!is_integer_type<T>::value && sizeof(CType) == 4)>
```
fixed to have the SumType be Decimal64 now

zeroshade commented on cpp/src/arrow/compute/kernels/hash_aggregate.cc:
```diff
@@ -892,8 +894,10 @@ struct GroupedVarStdImpl : public GroupedAggregator {
-  enable_if_t<is_floating_type<T>::value || (sizeof(CType) > 4), Status> ConsumeImpl(
-      const ExecSpan& batch) {
+  enable_if_t<is_floating_type<T>::value || (sizeof(CType) > 4) ||
+                  (!is_integer_type<T>::value && sizeof(CType) == 4),
```
fixed to have SumType for Decimal32 be Decimal64 now

zeroshade commented on cpp/src/arrow/compute/kernels/codegen_internal.h:
```diff
@@ -224,7 +258,9 @@ using enable_if_not_floating_value = enable_if_t<!std::is_floating_point<T>::val
-    enable_if_t<std::is_same<Decimal128, T>::value || std::is_same<Decimal256, T>::value,
+    enable_if_t<std::is_same<Decimal32, T>::value || std::is_same<Decimal64, T>::value ||
+                    std::is_same<Decimal128, T>::value ||
+                    std::is_same<Decimal256, T>::value,
```
In theory, because of how the templates work, as long as the value types have the proper methods implemented (operators and methods) then it *should* be safe, with the exception of dealing with the overflows that needed to be managed with the `GetSumType` and so on which we've already caught

felipecrv replied:
By safe I mean: all the things someone assumed when they wrote an `enable_if_decimal` still hold. And you're saying you checked all of them with `GetSumType` being the only one where assumptions didn't hold well.

bkietz replied:
It's a potentially breaking change for user code like:

```
template <typename T>
enable_if_decimal_value<T> SetToMagic(T* ptr) {
  constexpr char MAGIC[256] = ...;
  if constexpr (std::is_same_v<T, Decimal128>) {
    memcpy(ptr, &MAGIC, sizeof(Decimal128));
  } else {
    static_assert(std::is_same_v<T, Decimal256>); // error for Decimal32
    memcpy(ptr, &MAGIC, sizeof(MAGIC));
  }
}
```

I think that risk is outweighed by the greater weirdness of not including the new types however

pitrou replied:
This is an internal header, no user code should ever use anything from here.

pitrou commented:
Can we expedite the format change in a separate PR?

bkietz commented on cpp/src/arrow/engine/substrait/expression_internal.cc:
```diff
@@ -951,6 +951,38 @@ struct ScalarToProtoImpl {
   Status Visit(const MonthIntervalScalar& s) { return NotImplemented(s); }
   Status Visit(const DayTimeIntervalScalar& s) { return NotImplemented(s); }
 
+  Status Visit(const Decimal32Scalar& s) {
```
This is also a place where we'd benefit from a generic

zeroshade replied:
Converted to generic

bkietz commented on cpp/src/arrow/type.cc:
```diff
@@ -1459,6 +1478,42 @@ int32_t DecimalType::DecimalSize(int32_t precision) {
+
+Decimal32Type::Decimal32Type(int32_t precision, int32_t scale)
+    : DecimalType(type_id, 4, precision, scale) {
+  ARROW_CHECK_GE(precision, kMinPrecision);
+  ARROW_CHECK_LE(precision, kMaxPrecision);
```
Would it be worthwhile to extract

```
template <typename D>
Status ValidateDecimalPrecision(int32_t precision) {
  if (precision < D::kMinPrecision || precision > D::kMaxPrecision) {
    return Status::Invalid("Decimal precision out of range [", int32_t(D::kMinPrecision),
                           ", ", int32_t(D::kMaxPrecision), "]: ", precision);
  }
  return Status::Ok();
}
```
?

Then we could write
```suggestion
  ARROW_CHECK_OK(ValidatePrecisionScale<Decimal32Type>());
```

And in the `::Make` functions below
```
Result<std::shared_ptr<DataType>> Decimal32Type::Make(int32_t precision, int32_t scale) {
  RETURN_NOT_OK(ValidatePrecisionScale<Decimal32Type>(precision));
  return std::make_shared<Decimal32Type>(precision, scale);
}
```

zeroshade replied:
made these changes

zeroshade commented:
@pitrou I did that here: https://github.com/apache/arrow/pull/43976#pullrequestreview-2284369034

zeroshade commented on cpp/src/arrow/array/util.cc:
```diff
@@ -152,6 +152,42 @@ class ArrayDataEndianSwapper {
+      new_data[i] = bit_util::FromBigEndian(data[i]);
+#else
+      new_data[i] = bit_util::FromLittleEndian(data[i]);
+#endif
```
Made this change, let me know what you think

bkietz commented on cpp/src/arrow/array/builder_dict.h:
```diff
@@ -298,20 +298,12 @@ class DictionaryBuilderBase : public ArrayBuilder {
+    uint8_t data[T1::kByteWidth];
     value.ToBytes(data);
-    return Append(data, 32);
+    return Append(data, T1::kByteWidth);
```
Should be equivalent, little bit nicer:
```suggestion
    auto bytes = value.ToBytes();
    return Append(bytes.data(), bytes.size());
```

bkietz commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -31,6 +31,263 @@
+  /// \return error statis if the length is an invalid value
+  static Result<Decimal32> FromBigEndian(const uint8_t* data, int32_t length);
+
+  static void SwapEndian(const uint32_t* old_data, uint32_t* const new_data,
+                         const int64_t length);
```
This makes the usage in util.cc a bit nicer, but as a method on Decimal itself it might cause confusion. I was imagining
```suggestion
  Decimal32 SwapEndian() const;
```

Then util.cc would look like
```c++
  template <typename T>
  enable_if_decimal<T, Status> Visit(const T& type) {
    using value_type = typename TypeTraits<T>::CType;
    auto data = data_->buffers[1]->span_as<value_type>();
    auto new_data = new_buffer->mutable_data_as<value_type>();
    for (value_type v : data) {
      *new_data++ = v.SwapEndian();
    }
```

... actually now that I think of it, it'd also be pretty straightforward to just use ToBytes instead of adding SwapEndian:
```c++
  template <typename T>
  enable_if_decimal<T, Status> Visit(const T& type) {
    using value_type = typename TypeTraits<T>::CType;
    auto data = data_->buffers[1]->span_as<value_type>();
    auto new_data = new_buffer->mutable_data_as<value_type>();
    for (value_type v : data) {
      auto bytes = v.ToBytes();
      std::reverse(bytes.begin(), bytes.end());
      memcpy(new_data++, &bytes, sizeof(bytes));
    }
```

zeroshade replied:
updated using the suggestion, let me know what you think

bkietz reviewed (changes requested):
Thanks for taking the time to add a nicer generic decimal, looks great!

felipecrv commented on cpp/src/arrow/type_fwd.h:
```diff
@@ -512,9 +528,28 @@ std::shared_ptr<DataType> fixed_size_binary(int32_t byte_width);
 ///
 /// If the precision is greater than 38, a Decimal256Type is returned,
 /// otherwise a Decimal128Type.
+///
+/// Deprecated: prefer `smallest_decimal` instead.
```
`smallest_decimal` is soo0o long. `best_decimal`, `least_decimal`, `fast_decimal`...?

mapleFU commented on cpp/src/parquet/arrow/test_util.h:
```diff
@@ -142,7 +142,11 @@ template <int32_t byte_width>
+  if constexpr (byte_width == 4) {
+    decimals = gen.Decimal32(::arrow::decimal32(precision, 0), n);
+  } else if constexpr (byte_width == 8) {
+    decimals = gen.Decimal64(::arrow::decimal32(precision, 0), n);
```
why ::arrow::decimal32?

lidavidm replied:
presumably to disambiguate `parquet::arrow` and `arrow` namespaces

mapleFU replied:
I mean when bitwidth==8, why it's a decimal32 here 🤔?

lidavidm replied:
ah...good catch

mapleFU reviewed (commented):
We can leave an issue for Parquet native decimal32/decimal64 cc @wgtmac

bkietz commented on cpp/src/arrow/integration/json_internal.cc:
```diff
@@ -314,6 +314,20 @@ class SchemaWriter {
     writer_->Int(type.list_size());
   }
 
+  void WriteTypeMetadata(const Decimal32Type& type) {
```
I think we don't need to unpack to specific widths
```suggestion
  void WriteTypeMetadata(const DecimalType& type) {
```

bkietz commented on cpp/src/arrow/compare.cc:
```diff
@@ -750,6 +750,18 @@ class TypeEqualsVisitor {
     return Status::OK();
   }
 
+  Status Visit(const Decimal32Type& left) {
+    const auto& right = checked_cast<const Decimal32Type&>(right_);
+    result_ = left.precision() == right.precision() && left.scale() == right.scale();
+    return Status::OK();
+  }
```
```suggestion
  Status Visit(const DecimalType& left) {
    const auto& right = checked_cast<const DecimalType&>(right_);
    result_ = left.byte_width() == right.byte_width() && left.precision() == right.precision() && left.scale() == right.scale();
    return Status::OK();
  }
```

zeroshade commented:
I'll give until tomorrow afternoon. If there's no further comments / objections, i'll merge this.

Thanks everyone!

pitrou commented:
I plan to review this, could you wait for it?

zeroshade commented:
Sure, @pitrou I'll wait for your review

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_basic.cc:
```diff
@@ -336,8 +336,8 @@ struct ProductImpl : public ScalarAggregator {
-            this->product =
-                MultiplyTraits<AccType>::Multiply(*out_type, this->product, value);
+            this->product = MultiplyTraits<AccType>::Multiply(
+                *out_type, this->product, static_cast<ProductType>(value));
```
FTR, why was it necessary to add these casts?

zeroshade replied:
Because the constructors for Decimal32/Decimal64 are marked `explicit` I needed to explicitly add casts there to satisfy the templates for conversions

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_internal.h:
```diff
@@ -52,6 +52,16 @@ struct FindAccumulatorType<I, enable_if_floating_point<I>> {
 
+template <typename I>
+struct FindAccumulatorType<I, enable_if_decimal32<I>> {
+  using Type = Decimal64Type;
```
Why not `Decimal32Type`?

pitrou replied:
It doesn't make sense, especially if the decimal precision remains the same (which it probably does?).

pitrou replied:
By the way, the reason we upcast integers when summing them is to emulate what DB engines typically do. I don't think DB engines upcast decimals when summing them.

zeroshade replied:
That's a good point. we aren't changing the precision when we do this, so if it is necessary to avoid overflow, the type would need to be cast to increase the precision etc. I'll change it back to be the appropriate types

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_internal.h:
```diff
@@ -132,7 +142,13 @@ struct GetSumType<T, enable_if_integer<T>> {
 template <typename T>
-struct GetSumType<T, enable_if_decimal<T>> {
+struct GetSumType<T, enable_if_decimal32<T>> {
+  using SumType = arrow::Decimal64;
```
Why? This makes things slightly irregular and more difficult to predict for the user.

zeroshade replied:
See discussion [here](https://github.com/apache/arrow/pull/43957#discussion_r1750962486) it's to avoid overflows, we do the same for the sumtype for other integer types, which all have a Sum type of 64-bit integers

pitrou replied:
Decimal is **not** an integer type. The fact that it's coded as an integer is irrelevant.

zeroshade replied:
that's a good point and I agree, i'll put it back.

pitrou commented on cpp/src/arrow/integration/json_internal.cc:
```diff
@@ -969,12 +988,18 @@ Result<std::shared_ptr<DataType>> GetDecimal(const RjObject& json_type) {
   }
-  return Status::Invalid("Only 128 bit and 256 Decimals are supported. Received",
+
+  return Status::Invalid("Only 32/64/128/256-bit Decimals are supported. Received",
```
Let's add the required space at the end
```suggestion
  return Status::Invalid("Only 32/64/128/256-bit Decimals are supported. Received ",
```

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -343,8 +343,119 @@ struct DecimalGenerator {
+  std::shared_ptr<Array> MakeRandomArray(int64_t size, double null_probability,
+                                         int64_t alignment, MemoryPool* memory_pool) {
+    static constexpr int32_t kMaxDigitsInInteger = 9;
+    static constexpr int kNumIntegers = Decimal32Type::kByteWidth / 4;
+    static_assert(
+        kNumIntegers == (Decimal32Type::kMaxPrecision + kMaxDigitsInInteger - 1) /
+                            (kMaxDigitsInInteger + 1),
+        "inconsistent decimal metadata: kMaxPrecision doesn't match kByteWidth");
```
This can be massively simplified to:
```suggestion
    static_assert(kMaxDigitsInInteger >= Decimal32Type::kMaxPrecision);
```

zeroshade replied:
Isn't the purpose of this assert to validate that the precision *matches* `kByteWidth` as opposed to just verifying that the precision is equal to the max digits?

pitrou replied:
It's just a sanity check of the random generation algorithm.

zeroshade replied:
fair enough, simplified

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -343,8 +343,119 @@ struct DecimalGenerator {
+                            (kMaxDigitsInInteger + 1),
+        "inconsistent decimal metadata: kMaxPrecision doesn't match kByteWidth");
+
+    // First generate separate random values for individual components:
+    // boolean sign (including null-ness), and uint64 "digits" in big endian order.
```
This comment is outdated, remove it?

zeroshade replied:
you're right, that comment is for the decimal64/128/256 not for 32, i'll remove it

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -343,8 +343,119 @@ struct DecimalGenerator {
+    // boolean sign (including null-ness), and uint64 "digits" in big endian order.
+    const auto& decimal_type = checked_cast<const DecimalType&>(*type_);
+
+    auto remaining_digits = decimal_type.precision();
```
Why "remaining"?

zeroshade replied:
inherited from the original implementation for decimal128/256, i'll rename the variable to `digits_to_generate`

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -343,8 +343,119 @@ struct DecimalGenerator {
+};
+
+template <>
+struct DecimalGenerator<Decimal64Type> {
```
Hmm, can you perhaps refactor those two specialisations for `Decimal32Type` and `Decimal64Type` into a single one (perhaps `SmallDecimalGenerator<T>`)?

zeroshade replied:
I was able to create a single generic `SmallDecimalGenerator<T>` so this worked out :)

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -480,7 +494,7 @@ Result<std::shared_ptr<DataType>> MaybeMergeNumericTypes(
                           MaxDecimalDigitsForInteger(other_type->id()));
     ARROW_ASSIGN_OR_RAISE(const auto promoted_decimal,
-                          DecimalType::Make(promoted_type->id(), precision, 0));
+                          DecimalType::Make(promoted_type->id(), precision - 1, 0));
```
Why this change?

zeroshade replied:
The values we use for `MaxDecimalDigitsForInteger` is actually 1 + numeric_limits<T>::digits10 (i.e. for int32/uint32 we return 10 instead of 9, for int64 we return 19 instead of 18). As such we were passing a precision that was higher than the maximum precision allowed for Decimal32 and Decimal64 here (9 and 18 respectively).

it felt safer to me to put the -1 here than to change the value returned in `MaxDecimalDigitsForInteger` (which is used elsewhere and caused other complications in tests)

pitrou replied:
Ah! Got it.

pitrou commented on cpp/src/arrow/type.cc:
```diff
@@ -1459,20 +1478,51 @@ int32_t DecimalType::DecimalSize(int32_t precision) {
   return static_cast<int32_t>(std::ceil((precision / 8.0) * std::log2(10) + 1));
 }
 
+template <typename D>
+Status ValidateDecimalPrecision(int32_t precision) {
+  if (precision < D::kMinPrecision || precision > D::kMaxPrecision) {
+    return Status::Invalid("Decimal precision out of range [", int32_t(D::kMinPrecision),
+                           ", ", int32_t(D::kMaxPrecision), "]: ", precision);
+  }
+  return Status::OK();
+}
+
```
Make this `static` or anonymous?

pitrou commented on cpp/src/arrow/type_test.cc:
```diff
@@ -1241,15 +1245,15 @@ TEST_F(TestUnifySchemas, Decimal) {
-  CheckPromoteTo(int32(), decimal128(38, 37), decimal256(47, 37), options);
+  CheckPromoteTo(int32(), decimal128(38, 37), decimal256(46, 37), options);
   CheckUnifyFailsInvalid(decimal128(38, 10), decimal256(76, 5), options);
 
```
Can you add a couple tests that `smallest_decimal` returns the right type for a given precision?

zeroshade replied:
it gets tested in `array_test.cc` (though indirectly). i'll add some explicit tests for it

zeroshade replied:
tests added

pitrou commented on cpp/src/arrow/util/basic_decimal.h:
```diff
@@ -166,6 +167,317 @@ class ARROW_EXPORT GenericBasicDecimal {
 };
 
+template <typename BaseType>
+class ARROW_EXPORT BasicDecimal {
```
There's already a ton of "basic" types here, can you name it perhaps `SmallBasicDecimal`?

zeroshade replied:
renamed

pitrou commented on cpp/src/arrow/util/basic_decimal.h:
```diff
@@ -166,6 +167,317 @@ class ARROW_EXPORT GenericBasicDecimal {
   }
 };
 
+template <typename BaseType>
```
`BaseType` sounds like a base class, perhaps call it `DigitType` or simply `Digit`.

zeroshade replied:
renamed

pitrou commented on cpp/src/arrow/util/basic_decimal.h:
```diff
@@ -166,6 +167,317 @@ class ARROW_EXPORT GenericBasicDecimal {
+
+  BaseType value() const { return value_; }
+
+  /// \brief count hte number of leading binary zeroes.
```
```suggestion
  /// \brief count the number of leading binary zeroes.
```

pitrou commented on cpp/src/arrow/util/basic_decimal.h:
```diff
@@ -166,6 +167,317 @@ class ARROW_EXPORT GenericBasicDecimal {
+  using BasicDecimal<int32_t>::BasicDecimal;
+  using ValueType = int32_t;
+
+  /// \brief Negate the current value (in-place)
+  BasicDecimal32& Negate();
+  /// \brief Absolute value (in-place)
+  BasicDecimal32& Abs();
+  /// \brief Absolute value
+  static BasicDecimal32 Abs(const BasicDecimal32& left);
+  /// \brief Add a number to this one. The result is truncated to 32 bits.
+  BasicDecimal32& operator+=(const BasicDecimal32& right);
+  /// \brief Subtract a number from this one. The result is truncated to 32 bits.
+  BasicDecimal32& operator-=(const BasicDecimal32& right);
+  /// \brief Multiply this number by another. The result is truncated to 32 bits.
+  BasicDecimal32& operator*=(const BasicDecimal32& right);
```
1. why not define those in the base generic class?
2. since the implementation of each method is very short, perhaps inline it in the header here?

zeroshade replied:
I was following the pattern that was there for decimal128 and 256. I'll try inlining these

zeroshade replied:
these can't be defined in the base generic class because they return themselves by reference (you can't initialize a reference of the derived type from the base type)

I'll try just defining them inline in the header

pitrou replied:
You could use [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern) :-)

pitrou commented on cpp/src/arrow/util/basic_decimal.h:
```diff
@@ -166,6 +167,317 @@ class ARROW_EXPORT GenericBasicDecimal {
+  explicit operator BasicDecimal64() const;
+};
+
+ARROW_EXPORT bool operator<(const BasicDecimal32& left, const BasicDecimal32& right);
+ARROW_EXPORT bool operator<=(const BasicDecimal32& left, const BasicDecimal32& right);
+ARROW_EXPORT bool operator>(const BasicDecimal32& left, const BasicDecimal32& right);
+ARROW_EXPORT bool operator>=(const BasicDecimal32& left, const BasicDecimal32& right);
+
+ARROW_EXPORT BasicDecimal32 operator-(const BasicDecimal32& self);
+ARROW_EXPORT BasicDecimal32 operator+(const BasicDecimal32& left,
+                                      const BasicDecimal32& right);
+ARROW_EXPORT BasicDecimal32 operator-(const BasicDecimal32& left,
+                                      const BasicDecimal32& right);
+ARROW_EXPORT BasicDecimal32 operator*(const BasicDecimal32& left,
+                                      const BasicDecimal32& right);
```
Similarly, these ones can probably be 1) made generic 2) be inlined in this header

zeroshade replied:
If i try making them generic, it prevents implicit casting for these operators, causing compiler errors for things like `val < 0` as it can't work out whether to cast both to Decimal32/Decimal64 or cast both to int/int64 to determine which operator to call.

I also had to move these above their respective class definitions so that the inlined versions of `Abs` / `Negate` etc could use the `<` operator in them.

pitrou commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -31,6 +31,257 @@
+    constexpr auto min_value = std::numeric_limits<T>::min();
+    constexpr auto max_value = std::numeric_limits<T>::max();
+    const auto& self = *this;
+    if (self < min_value || self > max_value) {
+      return Status::Invalid("Invalid cast from Decimal32 value to ", sizeof(T),
+                             "byte integer");
+    }
```
Can this even happen?

zeroshade replied:
you're right, this can't happen for decimal32

pitrou commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -31,6 +31,257 @@
+    constexpr auto min_value = std::numeric_limits<T>::min();
+    constexpr auto max_value = std::numeric_limits<T>::max();
+    const auto& self = *this;
+    if (self < min_value || self > max_value) {
```
This will issue pointless comparisons if T is `int64_t`.

zeroshade replied:
you're right, these checks aren't needed for decimal64 or decimal32

pitrou commented on docs/source/status.rst:
```diff
@@ -44,6 +44,10 @@ Data Types
 +-------------------+-------+-------+-------+----+-------+-------+-------+-------+-----------+
 | Float32/64        | ✓     | ✓     | ✓     | ✓  |  ✓    |  ✓    | ✓     | ✓     | ✓         |
 +-------------------+-------+-------+-------+----+-------+-------+-------+-------+-----------+
+| Decimal32         | ✓     |       | ✓     |    |       |       |       |       |           |
++-------------------+-------+-------+-------+----+-------+-------+-------+-------+-----------+
+| Decimal64         | ✓     |       | ✓     |    |       |       |       |       |           |
++-------------------+-------+-------+-------+----+-------+-------+-------+-------+-----------+
```
Go isn't part of this PR, is it?

zeroshade replied:
the Go impl was already merged in the arrow-go repo

pitrou reviewed (changes requested):
Thanks a lot @zeroshade !

pitrou commented on cpp/src/arrow/compute/kernels/aggregate_internal.h:
```diff
@@ -52,6 +52,16 @@ struct FindAccumulatorType<I, enable_if_floating_point<I>> {
 };
 
+template <typename I>
+struct FindAccumulatorType<I, enable_if_decimal32<I>> {
```
Speaking of which, are there any tests for summing decimal32 and decimal64?

zeroshade replied:
Not yet, I was going to cover that when I do the rest of the Kernel additions for decimal32/decimal64 in a follow-up PR.

These were only added here to make everything compile due to adding the new types to the generalized `enable_if_decimal...` template helpers

pitrou commented on cpp/src/arrow/testing/random.cc:
```diff
@@ -343,8 +343,72 @@ struct DecimalGenerator {
+        rng_->Numeric<IntArrowType>(size, -1 * MaxDecimalInteger(digits_to_generate),
+                                    MaxDecimalInteger(digits_to_generate),
+                                    null_probability, alignment, memory_pool));
+
```
Note that we could just reuse the buffers, or perhaps even return `values.View(decimal_type)`.

zeroshade replied:
good point. updated

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -521,6 +726,23 @@ std::string Decimal128::ToString(int32_t scale) const {
   return str;
 }
 
+static inline void ShiftAndAdd(std::string_view input, uint32_t* out) {
```
This looks like a copy-paste job but I don't understand why it's useful. Converting a string to decimal32 or decimal64 shouldn't require multiple passes, so this should definitely be simplified.

zeroshade replied:
simplified this for decimal32 (decimal64 is able to just use the existing implementation used for the other types by just using a size of 1)

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -697,8 +919,150 @@ Status DecimalFromString(const char* type_name, std::string_view s, Decimal* out
+  }
+
+  if (out != nullptr) {
+    if constexpr (std::is_same_v<Decimal32, DecimalClass>) {
```
Basically, we should just be computing `ParseValue<uint64_t>(dec.whole_digits) * (10^dec.exponent) + ParseValue<uint64_t>(dec.fractional_digits)`.

zeroshade replied:
the `ShiftAndAdd` function handles the math for shifting `dec.whole_digits` before adding `dec.fractional_digits`

but if we want to just do that here explicitly rather than having a 32-bit overload for `ShiftAndAdd` i can do that.

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -342,6 +477,56 @@ struct Decimal128RealConversion
+  return Decimal32RealConversion::ToReal<double>(*this, scale);
+}
+
+std::string Decimal32::ToIntegerString() const { return std::to_string(value_); }
```
We should be using `FormatChars` to avoid locale-specific behavior.

zeroshade replied:
updated to use `internal::StringFormatter` like the other decimal types

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -342,6 +477,56 @@ struct Decimal128RealConversion
+  return Decimal64RealConversion::ToReal<double>(*this, scale);
+}
+
+std::string Decimal64::ToIntegerString() const { return std::to_string(value_); }
```
Also `FormatChars` here.

zeroshade replied:
updated to use `internal::StringFormatter` like the other decimal types

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
+  memcpy(reinterpret_cast<uint8_t*>(&result) + kMaxDecimalBytes - length, bytes, length);
+
+  const auto value = bit_util::FromBigEndian(result);
+  const int32_t bits_offset = std::max(0, length - kMaxDecimalBytes);
```
We guard against `length > kMaxDecimalBytes` above, so this is always 0?

zeroshade replied:
I can get rid of the `max` call and just use `length - kMaxDecimalBytes` then as this would always be `length - kMaxDecimalBytes`

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
+    return Decimal32(value);
+  }
+
+  int32_t final_value = -1 * (is_negative && length < kMaxDecimalBytes);
+  final_value = SafeLeftShift(final_value, bits_offset * CHAR_BIT);
+  final_value |= value;
```
This looks... complicated?

zeroshade replied:
I adapted this from the `Decimal128::FromBigEndian` but now that I'm re-reading it, am I wrong to think that I could just do `FromBigEndian` on the int32_t? (not `uint32_t`)

zeroshade replied:
Ah, I can't just blindly do `FromBigEndian` because i'm not guaranteed that `length` is exactly 4 bytes, hence why this shifting logic is needed for negative values.

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
+
+  const bool is_negative = static_cast<int8_t>(bytes[0]) < 0;
+
+  uint32_t result = 0;
```
How about something like:
```suggestion
  int32_t result = is_negative ? 0xffffffff : 0;
```
which should automically get you the required sign extension?

zeroshade replied:
wouldn't that get lost by the `memcpy` right below it?

zeroshade replied:
Ah! it would only get lost if `kMaxDecimalBytes == length` in which case we were just copying all the bytes anyways and the sign extension would already be there. Good point!

pitrou commented on cpp/src/arrow/util/decimal.h:
```diff
@@ -31,6 +31,243 @@
+  }
+
+  /// \brief Convert the Decimal32 value to a base 10 decimal string with the given scale
+  std::string ToString(int32_t scale) const;
```
It would be nice to add tests for the non-trivial methods: `ToString`, `FromString`, `ToReal`, `FromReal`, `FromBigEndian`.

(same for `Decimal64`)

zeroshade replied:
tests added

zeroshade commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -742,6 +1106,78 @@ static inline uint64_t UInt64FromBigEndian(const uint8_t* bytes, int32_t length)
+
+  const auto value = bit_util::FromBigEndian(result);
+  const int32_t bits_offset = std::max(0, length - kMaxDecimalBytes);
+  if (bits_offset == 8) {
```
you're right, this should be 4 not 8 so that we avoid an undefined shift by 32 below this

zeroshade replied:
i've added some comments to better explain this

zeroshade replied:
I've simplified this now

zeroshade commented:
@pitrou can you take another pass here? I believe I've addressed all of your comments. Thanks!

pitrou commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -1230,65 +1275,72 @@ TYPED_TEST(TestDecimalToRealDouble, LargeValues) {
 #endif  // __MINGW32__
 
-TEST(Decimal128Test, TestFromBigEndian) {
+TEST(Decimal32Test, TestFromBigEndian) {
```
It seems a number of tests are mostly or exactly identical between the Decimal tests, perhaps you can write a generic version and call it for each concrete test type?

zeroshade replied:
added generic versions for most of them. the LeftShift/RightShift and Negate tests are sufficiently different in their values that it's harder to make them generic

pitrou reviewed (commented):
Some more comments. Thanks for the updates!

zeroshade commented on cpp/src/arrow/util/decimal_internal.h:
```diff
@@ -442,8 +516,9 @@ struct RealTraits<float> {
-  // ceil(log10(2 ^ kMantissaBits))
-  static constexpr int kMantissaDigits = 8;
+  // log10(2 ^ kMantissaBits) ~= 7.2, let's be conservative to ensure more accuracy
+  // with our conversions for Decimal values
```
Yea, we had a persistent off-by-one rounding issue with Decimal32 and Decimal64 for the `FromReal` tests which was fixed by this, and had no detrimental effect on Decimal128/256

pitrou replied:
As I mentioned elsewhere, float->Decimal32 probably needs to fall back on the approx algorithm.

I'm a bit surprised for Decimal64, though. Is the RoundedRightShift right?

zeroshade replied:
I reran the tests to confirm, and yea it's only the Decimal32 case which runs into this problem (and only for Float since the double case falls back to the approx algorithm). 

In the float case, Decimal32 won't hit the condition to fall back on the approx algorithm as MaxPrecision == 9 and MantissaDigits is 8 (or 7 if I keep this change). In either case, using 7 for the mantissa digits is likely more accurate given that log10(2 ^ 24) is ~7.2 as opposed to doubles which uses 16 because log(2 ^ 53) ~= 15.9 so it makes sense to ceiling that to 16.

The only alternative would be to just have Decimal32 *always* go to the approx algorithm even for the float case despite maxprecision being larger than the mantissa digits.

zeroshade replied:
I've switched this back to 8 and just unconditionally send Decimal32 to the approx algorithm. Let me know if this is sufficient or if I should put it back.

zeroshade commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -109,6 +110,11 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
 
+    if constexpr (std::is_base_of_v<BasicDecimal32, DecimalType> &&
+                  std::is_same_v<Real, double>) {
+      return DecimalType::FromReal(static_cast<float>(real), precision, scale);
```
the example you give would essentially be the same condition (though more concise) which would always affect double->Decimal32 conversions. That said, using `FromPositiveRealApprox` is likely better than a blanket drop to float. I'll give this a try and make sure the tests still pass for it.

pitrou replied:
It's the right condition logically, though :-)

zeroshade replied:
Also that haha

zeroshade commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -156,7 +162,9 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
         // lose the last digit. If `precision` is almost the full precision,
         // there can be an off-by-one error due to rounding.
-        const int mul_step = std::max(1, kMaxPrecision - precision);
+        constexpr int is_dec32_or_dec64 =
+            DecimalType::kByteWidth <= BasicDecimal64::kByteWidth;
+        const int mul_step = std::max(1, kMaxPrecision - precision - is_dec32_or_dec64);
```
With decimal32 and decimal64 I needed to reduce the `mul_step` by 1 in order to eliminate off-by-one rounding errors. I didn't want to unconditionally add extra operations to decimal128/256 by lowering their `mul_step` so I only did it for decimal32/64

pitrou replied:
That's several things that need to be arbitrarily lowered, and sounds a bit unexpected.
Random hacks like this are a bad smell, especially if there's no detailed explanation other that "it works better".

pitrou replied:
That said, if you don't want to debug these now, I would favor removing the hacks, adding temporary workarounds in the tests, and opening an issue together with value(s) reproducing the problem. Then this can be investigated (by you, me or @bkietz for example :-)).

zeroshade replied:
I've adjusted the tests to simply not hit this case anymore with the off-by-one rounding issue for decimal64. 

Let me know if you prefer that to the above hack with explanation.

zeroshade commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -1170,40 +1206,49 @@ TYPED_TEST(TestDecimalToRealFloat, LargeValues) {
 }
 
 TYPED_TEST(TestDecimalToRealFloat, Precision) {
```
ah, i missed that one!

zeroshade commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -896,22 +917,30 @@ TYPED_TEST_SUITE(TestDecimalFromRealFloat, DecimalTypes);
 
 TYPED_TEST(TestDecimalFromRealFloat, SuccessConversion) {
```
done

zeroshade commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -1389,6 +1435,838 @@ std::vector<CType> GetRandomNumbers(int32_t size) {
+  ASSERT_OK(whole.ToInteger(&out));
+  ASSERT_EQ(123456, out);
+  ASSERT_OK(fraction.ToInteger(&out));
+  ASSERT_EQ(0, out);
```
added a helper via lambda like in other tests

zeroshade commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -1389,6 +1435,838 @@ std::vector<CType> GetRandomNumbers(int32_t size) {
+  Decimal32 result;
+  int32_t out;
+
+  result = Decimal32("1234").IncreaseScaleBy(0);
```
done

pitrou commented on cpp/src/arrow/util/decimal_test.cc:
```diff
@@ -1336,43 +1406,42 @@ TEST(Decimal128Test, TestFromBigEndian) {
-  Decimal128 value1("1234");
-  int32_t out1;
+TYPED_TEST(TestBasicDecimalFunctionality, TestToInteger) {
+  if constexpr (std::is_same_v<TypeParam, Decimal256>) {
```
`Decimal256` isn't in `BasicFunctionalityDecimalTypes`, so is this condition useful? Or were you planning to add back `Decimal256` into the type list?

zeroshade replied:
that's a holdover from before I decided not to add it to the `BasicFunctionalityDecimalTypes`, i'll remove it.

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -92,6 +93,12 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
     constexpr int kMantissaDigits = RealTraits<Real>::kMantissaDigits;
 
+    // to avoid precision and rounding issues, we'll unconditionally
+    // throw Decimal32 to the approx algorithm instead.
```
Can we open a GH issue to improve this and reference it here?
(for example, a very simple solution would be to convert to Decimal64 and convert down to Decimal32 afterwards)

zeroshade replied:
Issue opened and referenced here (#44216)

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -109,6 +116,10 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
       return OverflowError(real, precision, scale);
     }
 
+    if constexpr (kMaxPrecision <= kMantissaDigits) {
```
```suggestion
    // The algorithm below requires the destination decimal type
    // to be strictly more precise than the source float type
    // (see `kSafeMulByTenTo` calculation).
    if constexpr (kMaxPrecision <= kMantissaDigits) {
```

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -521,6 +750,23 @@ std::string Decimal128::ToString(int32_t scale) const {
+  tmp *= kUInt32PowersOfTen[len];
+  tmp += value;
+
+  *out = static_cast<uint32_t>(tmp & 0xFFFFFFFFU);
```
So we're just losing bits if there's an overflow here? I'll note that `SimpleDecimalFromString` checks for overflows on other ops.

How about:
1. remove the `uint32_t`-output variant of this function
2. for Decimal32, call the `uint64_t` variant and reuse the existing limits check

zeroshade replied:
`SimpleDecimalFromString` has the same overflow check for both Decimal32 and Decimal64 though? 

that said, removing the `uint32_t` variant and just using the `uint64_t` variant does simplify this, so I'll do that.

pitrou commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -697,8 +943,150 @@ Status DecimalFromString(const char* type_name, std::string_view s, Decimal* out
+      if (dec.sign == '-') {
+        out->Negate();
+      }
+    } else {
```
The two codepaths are almost identical except for `uint32_t` vs. `uint64_t`, you can probably reconcile them (perhaps something like `make_unsigned_t<typename DecimalClass::DigitType>`?)

zeroshade replied:
following the suggestion from https://github.com/apache/arrow/pull/43957#discussion_r1771711880, i was able to eliminate the branch there entirely. Now the `Decimal32` case just uses the `uint64_t` variant anyways, performs the overflow check and then constructs the decimal value from the `uint64_t` by downcasting. 

So the codepaths have been unified

pitrou reviewed (commented):
Thanks @zeroshade ! Some more comments.

zeroshade commented:
ping @pitrou for another round :) we're getting closer! haha

pitrou reviewed (approved):
+1 from me. Kudos for doing this and being persistent through reviews @zeroshade !

zeroshade commented on cpp/src/arrow/util/decimal.cc:
```diff
@@ -237,6 +251,140 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
+  using Base::LargePowerOfTen;
+  using Base::PowerOfTen;
+
+  static Decimal32 RoundedRightShift(const Decimal32& x, int bits) {
```
that's a fair point, i'll update the non-Approx methods for the decimal32 version to be no-ops + a failing `DCHECK`

zeroshade replied:
actually, we only throw over to Approx for the conversion *from* Real, we still use the other methods for `ToReal` so only the `RoundedRightShift` method needs to be a no-op. the other methods are still being used.
